### PR TITLE
Penultimate round of `db.DefaultContext` refactor

### PR DESCRIPTION
**Pull request by JakobDev**
Part of #27065

### CODE REVIEW

File: models/auth/source.go [hunk 1]
```diff
@@ -232,53 +233,53 @@ func CreateSource(source *Source) error {
...
+		if _, err := db.GetEngine(ctx).Delete(source); err != nil {
 			log.Error("CreateSource: Error while wrapOpenIDConnectInitializeError: %v", err)
 		}
 	}
 	return err
 }
 
 // Sources returns a slice of all login sources found in DB.
-func Sources() ([]*Source, error) {
+func Sources(ctx context.Context) ([]*Source, error) {
```
Comment (delvh) [inline models/auth/source.go:244]: I don't know if 
```go
admin/user/new.tmpl:19:                                                 {{range .Sources}}
admin/user/edit.tmpl:23:                                                        {{range .Sources}}
admin/auth/list.tmpl:23:                                        {{range .Sources}}
```
are affected by this change.

File: models/issues/issue_project.go [hunk 2]
```diff
@@ -38,8 +38,8 @@ func (issue *Issue) projectID(ctx context.Context) int64 {
 }
 
 // ProjectBoardID return project board id if issue was assigned to one
-func (issue *Issue) ProjectBoardID() int64 {
-	return issue.projectBoardID(db.DefaultContext)
+func (issue *Issue) ProjectBoardID(ctx context.Context) int64 {
+	return issue.projectBoardID(ctx)
 }
 
 func (issue *Issue) projectBoardID(ctx context.Context) int64 {
```
Comment (delvh) [inline models/issues/issue_project.go]:
```suggestion
```

File: templates/shared/issueicon.tmpl [hunk 3]
```diff
@@ -7,7 +7,7 @@
 		{{if .IsClosed}}
 			{{svg "octicon-git-pull-request" 16 "text red"}}
 		{{else}}
-			{{if and .PullRequest .PullRequest.IsWorkInProgress}}
+			{{if and .PullRequest (.PullRequest.IsWorkInProgress ctx)}}
 				{{svg "octicon-git-pull-request-draft" 16 "text grey"}}
 			{{else if and (.GetPullRequest ctx) (.GetPullRequest ctx).IsWorkInProgress}}
```
Comment (delvh) [inline templates/shared/issueicon.tmpl]: ctx missing
```suggestion
			{{else if and (.GetPullRequest ctx) ((.GetPullRequest ctx).IsWorkInProgress ctx)}}
```

File: templates/repo/view_list.tmpl [hunk 4]
```diff
@@ -26,7 +26,7 @@
 					</a>
 					{{template "repo/commit_statuses" dict "Status" .LatestCommitStatus "Statuses" .LatestCommitStatuses "root" $}}
 					{{$commitLink:= printf "%s/commit/%s" .RepoLink (PathEscape .LatestCommit.ID.String)}}
-					<span class="grey commit-summary" title="{{.LatestCommit.Summary}}"><span class="message-wrapper">{{RenderCommitMessageLinkSubject $.Context .LatestCommit.Message $.RepoLink $commitLink $.Repository.ComposeMetas}}</span>
+					<span class="grey commit-summary" title="{{.LatestCommit.Summary}}"><span class="message-wrapper">{{RenderCommitMessageLinkSubject $.Context .LatestCommit.Message $.RepoLink $commitLink ($.Repository.ComposeMetas ctx)}}</span>
 						{{if IsMultilineCommitMessage .LatestCommit.Message}}
 							<button class="ui button js-toggle-commit-body ellipsis-button" aria-expanded="false">...</button>
 							<pre class="commit-body gt-hidden">{{RenderCommitBody $.Context .LatestCommit.Message $.RepoLink $.Repository.ComposeMetas}}</pre>
```
Comment (delvh) [inline templates/repo/view_list.tmpl]:
```suggestion
							<pre class="commit-body gt-hidden">{{RenderCommitBody $.Context .LatestCommit.Message $.RepoLink ($.Repository.ComposeMetas ctx)}}</pre>
```

**Review by delvh — COMMENTED** [review]
Other instances I've found:
`ComposeMetas`:
```go
user/dashboard/feeds.tmpl
93:                                                                     {{RenderCommitMessage $.Context .Message $repoLink $.ComposeMetas}}

repo/commit_page.tmpl
22:                             <h3 class="gt-mb-0 gt-f1"><span class="commit-summary" title="{{.Commit.Summary}}">{{RenderCommitMessage $.Context .Commit.Message $.RepoLink $.Repository.ComposeMetas}}</span>{{template "repo/commit_statuses" dict "Status" .CommitStatus "Statuses" .CommitStatuses "root" $}}</h3>
138:                            <pre class="commit-body">{{RenderCommitBody $.Context .Commit.Message $.RepoLink $.Repository.ComposeMetas}}</pre>
279:                            <pre class="commit-body">{{RenderNote $.Context .Note $.RepoLink $.Repository.ComposeMetas}}</pre>

repo/branch/list.tmpl
28:                                                             <p class="info gt-df gt-ac gt-my-2">{{svg "octicon-git-commit" 16 "gt-mr-2"}}<a href="{{.RepoLink}}/commit/{{PathEscape .DefaultBranchBranch.DBBranch.CommitID}}">{{ShortSha .DefaultBranchBranch.DBBranch.CommitID}}</a> · <span class="commit-message">{{RenderCommitMessage $.Context .DefaultBranchBranch.DBBranch.CommitMessage .RepoLink .Repository.ComposeMetas}}</span> · {{ctx.Locale.Tr "org.repo_updated"}} {{TimeSince .DefaultBranchBranch.DBBranch.CommitTime.AsTime ctx.Locale}}{{if .DefaultBranchBranch.DBBranch.Pusher}} &nbsp;{{template "shared/user/avatarlink" dict "user" .DefaultBranchBranch.DBBranch.Pusher}}{{template "shared/user/namelink" .DefaultBranchBranch.DBBranch.Pusher}}{{end}}</p>
104:                                                                    <p class="info gt-df gt-ac gt-my-2">{{svg "octicon-git-commit" 16 "gt-mr-2"}}<a href="{{$.RepoLink}}/commit/{{PathEscape .DBBranch.CommitID}}">{{ShortSha .DBBranch.CommitID}}</a> · <span class="commit-message">{{RenderCommitMessage $.Context .DBBranch.CommitMessage $.RepoLink $.Repository.ComposeMetas}}</span> · {{ctx.Locale.Tr "org.repo_updated"}} {{TimeSince .DBBranch.CommitTime.AsTime ctx.Locale}}{{if .DBBranch.Pusher}} &nbsp;{{template "shared/user/avatarlink" dict "user" .DBBranch.Pusher}} &nbsp;{{template "shared/user/namelink" .DBBranch.Pusher}}{{end}}</p>

repo/commits_list_small.tmpl
41:             <span class="gt-mono commit-summary {{if gt .ParentCount 1}} grey text{{end}}" title="{{.Summary}}">{{RenderCommitMessageLinkSubject $.root.Context .Message ($.comment.Issue.PullRequest.BaseRepo.Link|Escape) $commitLink $.comment.Issue.PullRequest.BaseRepo.ComposeMetas}}</span>
46:                     <pre class="commit-body gt-hidden">{{RenderCommitBody $.root.Context .Message ($.comment.Issue.PullRequest.BaseRepo.Link|Escape) $.comment.Issue.PullRequest.BaseRepo.ComposeMetas}}</pre>

repo/graph/commits.tmpl
32:                                             <span>{{RenderCommitMessage $.Context $commit.Subject $.RepoLink $.Repository.ComposeMetas}}</span>

repo/commits_list.tmpl
63:                                                             <span class="commit-summary {{if gt .ParentCount 1}} grey text{{end}}" title="{{.Summary}}">{{RenderCommitMessageLinkSubject $.Context .Message $commitRepoLink $commitLink $.Repository.ComposeMetas}}</span>
71:                                                     <pre class="commit-body gt-hidden">{{RenderCommitBody $.Context .Message $commitRepoLink $.Repository.ComposeMetas}}</pre>

repo/diff/compare.tmpl
197:                                            <span id="issue-title">{{RenderIssueTitle $.Context .PullRequest.Issue.Title $.RepoLink $.Repository.ComposeMetas}}</span>
```
`AllowsPulls`:
```go

repo/branch/list.tmpl
128:                                                                            {{else if and (not .DBBranch.IsDeleted) $.AllowsPulls (gt .CommitsAhead 0)}}
134:                                                                            {{if and (not .DBBranch.IsDeleted) $.AllowsPulls (gt .CommitsAhead 0)}}
```

File: models/auth/source.go (hunk 1 above)
Comment (JakobDev) [reply models/auth/source.go:244]: Sources is passed through `ctx.Data` here. The pages render.

**Review by delvh — APPROVED** [review]

**Comment by delvh** [comment]
Ping for other reviewers

**Review by wxiaoguang — APPROVED** [review]
I do not feel safe to change the template-related code too much (there are many regressions).

In most (all) cases, the methods used by templates don't need to strictly follow the database context because they can't be in a transaction.

To (partially) resolve the "template code is dynamic" problem, using unique and longer names is the only choice. That's why I keep calling out to use at least 2 words for method/variable names. Then many `.Link` regressions could be avoided. And the changed functions should be searched in all templates when refactoring.

And to make the release progress safe, the refactoring time is also important. It's better to do it when the last release is stable enough, and avoid backporting large refactoring. Then people would have enough time to eat the dogfood.

----

Since the refactoring has started, it shouldn't stop halfway, so I vote my approval here. Hopefully the regressions could be fixed ASAP.

**Comment by silverwind** [comment]
Dashboard does not render any more for me. Is it related?

```
Render failed, failed to render template: user/dashboard/dashboard, error: template error: builtin(static):user/dashboard/feeds:93:62 : executing "user/dashboard/feeds" at <$.ComposeMetas>: ComposeMetas is not a method but has arguments
----------------------------------------------------------------------
									{{RenderCommitMessage $.Context .Message $repoLink ($.ComposeMetas ctx)}}
```

**Comment by JakobDev** [comment]
> Is it related?

It is related. This variable has the same name as a function I changed.

> I do not feel safe to change the template-related code too much

I'm also not sure about it, but the current refactoring don't land in 1.21, so we have enough time to find problems.

> In most (all) cases, the methods used by templates don't need to strictly follow the database context because they can't be in a transaction.

I recently had a database deadlock, because one function called in a transaction was using `db.DefaultContext`. This function was not used by a template, but there is always a risk, so I think it's better to take the pill and change it during the refactoring.

Luckily the next round will likely be the last.

**Comment by silverwind** [comment]
We need at least some test coverage for typical pages which have been recently regressed with these refactors, like frontpage, pull request view etc. Should be complex scenarios with all the states in them etc.

**Pull request merged**